src/Symfony/Component/String/GraphemeString.php

```diff
@@ -0,0 +1,340 @@
+
+        return $str;
+    }
+
+    public function replace(string $from, string $to): parent
+    {
+        if (false === $from = normalizer_normalize($from)) {
+            throw new InvalidArgumentException('Invalid UTF-8 string.');
+        }
+
+        if ('' !== $to && !preg_match('//u', $to)) {
+            throw new InvalidArgumentException('Invalid UTF-8 string.');
+        }
+
+        $str = clone $this;
+
+        if ('' !== $from) {
+            if ($this->ignoreCase) {
+                $str->string = normalizer_normalize(implode($to, preg_split('{'.preg_quote($from).'}iu', $this->string)));
+            } else {
+                $str->string = normalizer_normalize(str_replace($from, $to, $this->string));
+            }
+        }
+
+        return $str;
+    }
+
+    public function replaceMatches(string $fromPattern, $to): parent
+    {
+        $str = parent::replaceMatches($fromPattern, $to);
+        $str->string = normalizer_normalize($string);
+
+        return $str;
+    }
+
+    public function slice(int $start = 0, int $length = null): parent
+    {
+        $str = clone $this;
+        $str->string = (string) grapheme_substr($this->string, $start, $length ?? PHP_INT_MAX);
+
+        return $str;
+    }
+
+    public function splice(string $replacement, int $start = 0, int $length = null): parent
+    {
+        if (!preg_match('//u', $replacement)) {
+            throw new InvalidArgumentException('Invalid UTF-8 string.');
+        }
+
+        $str = clone $this;
+        $start = $start ? \strlen(grapheme_substr($this->string, 0, $start)) : 0;
+        $length = $length ? \strlen(grapheme_substr($this->string, $start, $length ?? PHP_INT_MAX)) : $length;
+        $str->string = substr_replace($this->string, $replacement, $start, $length ?? PHP_INT_MAX);
+        $str->string = normalizer_normalize($this->string);
+
+        return $str;
+    }
+
+    public function split(string $delimiter, int $limit = null): array
+    {
+        if ('' === $delimiter) {
+            return $this->chunk();
+        }
+
+        if (false === $delimiter = normalizer_normalize($delimiter)) {
+            throw new InvalidArgumentException('Invalid UTF-8 string.');
+        }
+
+        $chunks = $this->ignoreCase
+            ? preg_split('{'.preg_quote($delimiter).'}iu', $this->string, $limit ?? PHP_INT_MAX)
+            : explode($delimiter, $this->string, $limit ?? PHP_INT_MAX);
+
+        foreach ($chunks as $i => $chunk) {
+            $chunks[$i] = $str = clone $this;
+            $str->string = $chunk;
+        }
+
+        return $chunks;
+    }
+
+    public function startsWith(string $prefix): bool
+    {
+        if ('' === $prefix) {
+            return true;
+        }
+
+        if (false === $prefix = normalizer_normalize($prefix)) {
+            return false;
+        }
+
+        return 0 === ($this->ignoreCase ? grapheme_stripos($this->string, $prefix) : strpos($this->string, $prefix));
```

Comment:
Do we need `grapheme_strpos` here, instead of `strpos`?
https://3v4l.org/GGSBu

Reply:
nice catch, implementations updated.